cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"log"
+	"reflect"
+	"sort"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/pkg/transport"
+)
+
+// EtcdClusterInterrogator is an interface to get etcd cluster related information
+type EtcdClusterInterrogator interface {
+	GetEtcdClusterStatus() (*EtcdStatus, error)
+}
+
+// EtcdCluster provides connection parameters for an etcd cluster
+type EtcdCluster struct {
+	Endpoints []string
+	TLS       *tls.Config
+}
+
+// EtcdStatusChecker an interface for etcd status validation functions
+type EtcdStatusChecker interface {
+	VersionsMatch() bool
+}
+
+// EtcdStatus provides data about an etcd cluster
+type EtcdStatus struct {
+	Versions []string
+	Members  []uint64
```

Comment:
I think you could simplify the logic and extra functions by just having map[int]string

Reply:
I have added it in this manner because I think there may be a handful of other EtcdStatus interface functions that might be handy (ie. determining whether the version of etcd is at a version that is appropriate). As Members is not really being used yet, maybe we ditch that field?